### PR TITLE
Census tract support

### PR DESCRIPTION
**Pull request by kindofluke**
Adding support for census tracts via the CLI. 

There are a few issues with this branch that may prevent merging. 

1. All tests are passing but we don't have great tract level test data. We are working to run the original STATA data to get some valid testing. 
2. the new CLI options are somewhat verbose
3. the GUI frontend does not support tracts as yet.

### CODE REVIEW

**Comment by theonaunheim** [comment]
Many thanks, @kindofluke .

I will review this weekend and start working through the PR.

Housekeeping item: can you please confirm you are OK licensing this contribution to surgeo via the MIT license? I am OK with you keeping copyright to your code and will add you to license/docs ... but I like to keep it under a single license for downstream users.

**Comment by kindofluke** [comment]
Sent via email

File: surgeo/app/surgeo_cli.py [hunk 1]
```diff
@@ -119,13 +131,27 @@ def _load_df(self):
 
     def _run_geo(self, df):
         """Method called from self._process_df() to get geo results"""
-        model = GeocodeModel()
+        if self._ct:
+            model = GeocodeModel("TRACT")
+        else:
+            model = GeocodeModel("ZCTA")
+        # If an optional name is speicied, select that column and run
```
Comment (TheCleric) [inline surgeo/app/surgeo_cli.py:138]:
```suggestion
        # If an optional name is specified, select that column and run
```

File: README.rst [hunk 2]
```diff
@@ -110,7 +110,7 @@ To use the CLI, type in "surgeo" followed by your arguments.
 As a Module
 ~~~~~~~~~~~
 
-Surgeo is best used as a module.
+Surgeo is best used as a module. Add Census Tract Input Examples.
```
Comment (TheCleric) [inline README.rst:113]: Is this like a TODO comment?

File: surgeo/app/surgeo_cli.py [hunk 3]
```diff
@@ -119,13 +131,27 @@ def _load_df(self):
...
             result = model.get_probabilities(target)
         # Otherwise use 'zcta5' (and raise error if need be.)
+        elif self._state_col is not None and self._ct:
+            target = df[[self._state_col, self._county_col, self._tract_col]]
+            result = model.get_probabilities_tract(target)
+        elif self._ct:
+            try:
+                target = df[['state', 'column', 'tract']]
+            except KeyError:
+                raise SurgeoException("No state county or tract column found")
```
Comment (TheCleric) [inline surgeo/app/surgeo_cli.py:154]:
```suggestion
                raise SurgeoException("Columns for state, county, and tract not found.")
```

File: surgeo/models/geocode_model.py [hunk 4]
```diff
@@ -30,9 +30,12 @@ class GeocodeModel(BaseModel):
 
     """
 
-    def __init__(self):
+    def __init__(self, geo_level='ZCTA'):
         super().__init__()
-        self._PROB_RACE_GIVEN_ZCTA = self._get_prob_race_given_zcta()
+        if geo_level == 'TRACT':
```
Comment (TheCleric) [inline surgeo/models/geocode_model.py:35]:
```suggestion
        if geo_level.upper() == 'TRACT':
```

File: surgeo/models/surgeo_model.py [hunk 5]
```diff
@@ -67,12 +68,16 @@ class SurgeoModel(BaseModel):
         69. `<https://link.springer.com/article/10.1007/s10742-009-0047-1>`_
 
     """
-    def __init__(self):
+    def __init__(self, geo_level="ZCTA"):
         super().__init__()
-        self._PROB_ZCTA_GIVEN_RACE = self._get_prob_zcta_given_race()
+        self.geo_level = geo_level
+        if geo_level == "TRACT":
```
Comment (TheCleric) [inline surgeo/models/surgeo_model.py:74]:
```suggestion
        if self.geo_level == "TRACT":
```

File: surgeo/models/surgeo_model.py [hunk 6]
```diff
@@ -88,8 +93,8 @@ def get_probabilities(self, names, zctas):
         ----------
         names : pd.Series
             A series of names to use for the BISG algorithm
-        zctas : pd.Series
-            A series of ZIP/ZCTA codes for the BISG algorithm
+        geo_df : Union[pd.Series , pd.DataFrame]
```
Comment (TheCleric) [inline surgeo/models/surgeo_model.py:96]:
```suggestion
        geo_df : Union[pd.Series, pd.DataFrame]
```

File: surgeo/models/surgeo_model.py [hunk 7]
```diff
@@ -166,18 +176,29 @@ def _get_surname_probs(self,
         )
         return surname_probs
 
-    def _get_geocode_probs(self, zctas: pd.Series) -> pd.DataFrame:
+    def _get_geocode_probs(self, geo_df: Union[pd.Series,pd.DataFrame]) -> pd.DataFrame:
```
Comment (TheCleric) [inline surgeo/models/surgeo_model.py:179]:
```suggestion
    def _get_geocode_probs(self, geo_df: Union[pd.Series, pd.DataFrame]) -> pd.DataFrame:
```

File: surgeo/models/surgeo_model.py [hunk 8]
```diff
@@ -67,12 +68,16 @@ class SurgeoModel(BaseModel):
         69. `<https://link.springer.com/article/10.1007/s10742-009-0047-1>`_
 
     """
-    def __init__(self):
+    def __init__(self, geo_level="ZCTA"):
         super().__init__()
-        self._PROB_ZCTA_GIVEN_RACE = self._get_prob_zcta_given_race()
+        self.geo_level = geo_level
```
Comment (TheCleric) [inline surgeo/models/surgeo_model.py:73]:
```suggestion
        self.geo_level = geo_level.upper()
```

**Comment by theonaunheim** [comment]
Good afternoon, @kindofluke and @TheCleric . Thanks for the contributions and apologies for the delay; I wasn't able to address these last weekend.

These look great. I've merged them to dev for testing. I anticipate that I should be able to complete that today. After that I will:

* accept the PR;
* merge the dev modifications to master;
* create the Windows EXE;
* package for PyPI; and,
* modify master with new links.

Should be able to wrap this up by Sunday evening CST.

**Comment by kindofluke** [comment]
> Good afternoon, @kindofluke and @TheCleric . Thanks for the contributions and apologies for the delay; I wasn't able to address these last weekend.
> 
> These look great. I've merged them to dev for testing. I anticipate that I should be able to complete that today. After that I will:
> 
>     * accept the PR;
> 
>     * merge the dev modifications to master;
> 
>     * create the Windows EXE;
> 
>     * package for PyPI; and,
> 
>     * modify master with new links.
> 
> 
> Should be able to wrap this up by Sunday evening CST.

Theo, one other note I wasn't able to integrate census tracts into the  tkinter app. Not sure i mentioned that.

**Comment by theonaunheim** [comment]
@kindofluke , understood. When I do the 2020 census rewrite I will shoot for parity with other models (probably by making separate SurTract and SurZcta models). For now adding this to the GUI is more trouble than it's worth.

**Pull request merged**